core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java

```diff
@@ -843,6 +669,274 @@ private void syncLocalToRemote() {
+   */
+  private void syncRemoteToLocal() {
+    logger.info(String.format(
+        Locale.US,
+        "t='%d': syncRemoteToLocal START",
+        logicalT));
+
+    for (final NamespaceSynchronizationConfig nsConfig : syncConfig) {
+      final MongoCollection<BsonDocument> localColl = getLocalCollection(nsConfig.getNamespace());
+      final Map<BsonValue, ChangeEvent<BsonDocument>> remoteChangeEvents =
+          instanceChangeStreamListener.getEventsForNamespace(nsConfig.getNamespace());
+
+      final Set<BsonValue> unseenIds = nsConfig.getSynchronizedDocumentIds();
+      final Set<BsonDocument> staleDocuments = getStaleDocuments(nsConfig);
+      final Set<BsonValue> staleIds = getStaleDocumentIds(staleDocuments);
+      final Map<BsonValue, BsonDocument> staleDocumentMap = new HashMap<>();
+      for (final BsonDocument staleDocument : staleDocuments) {
+        staleDocumentMap.put(staleDocument.get("_id"), staleDocument);
+      }
+
+      for (final Map.Entry<BsonValue, ChangeEvent<BsonDocument>> eventEntry :
+          remoteChangeEvents.entrySet()) {
+        logger.info(String.format(
+            Locale.US,
+            "t='%d': syncRemoteToLocal consuming event of type: %s",
+            logicalT,
+            eventEntry.getValue().getOperationType()));
+
+        final CoreDocumentSynchronizationConfig docConfig =
+            nsConfig.getSynchronizedDocument(eventEntry.getKey().asDocument().get("_id"));
+
+        if (docConfig == null) {
+          // Not interested in this event.
+          continue;
+        }
+
+        unseenIds.remove(docConfig.getDocumentId());
+        staleIds.remove(docConfig.getDocumentId());
+        syncRemoteChangeEventToLocal(nsConfig, docConfig, eventEntry.getValue(), localColl);
+      }
+
+      for (final BsonValue docId : unseenIds) {
+        final CoreDocumentSynchronizationConfig docConfig =
+            nsConfig.getSynchronizedDocument(docId);
+        if (docConfig == null) {
+          // means we aren't actually synchronizing on this remote doc
+          continue;
+        }
+
+        if (nsConfig.isStale() && staleIds.contains(docId)) {
+          if (docConfig.hasCommittedVersion(DocumentVersionInfo
+              .getRemoteVersionInfo(staleDocumentMap.get(docId)).getCurrentVersion())) {
+            // Skip this event since we generated it.
+            logger.info(String.format(
+                Locale.US,
+                "t='%d': syncRemoteChangeEventToLocal ns=%s documentId=%s remote change event was "
+                    + "generated by us; dropping the event",
+                logicalT,
+                nsConfig.getNamespace(),
+                docConfig.getDocumentId()));
+            continue;
+          }
+
+          syncRemoteChangeEventToLocal(
+              nsConfig,
+              docConfig,
+              changeEventForLocalReplace(
+                  nsConfig.getNamespace(),
+                  docId,
+                  staleDocumentMap.get(docId),
```

Comment:
This should come from the result of a remote query to get the latest docs. So there should be something that indicates what doc ids are stale and then something get the latest for those.

Reply:
👍